transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java

```diff
@@ -84,9 +84,9 @@ int process(IOUringCompletionQueueCallback callback) {
             PlatformDependent.putIntOrdered(kHeadAddress, ringHead);
 
             int fd = (int) (udata >>> 32);
-            int opMask = (int) (udata & 0xFFFFFFFFL);
+            int opMask = (int) udata;
             int op = opMask >>> 16;
-            int data = opMask & 0xffff;
+            int data = (short) opMask;
```

Comment:
sorry not sure I follow... Can you show me how you would change this ?

Reply:
```patch
Index: transport-native-io_uring/src/test/java/io/netty/channel/uring/NativeTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/test/java/io/netty/channel/uring/NativeTest.java	(revision ea5a539bc60ad29d40895844051c61e482338cc0)
+++ transport-native-io_uring/src/test/java/io/netty/channel/uring/NativeTest.java	(date 1601286542688)
@@ -59,7 +59,7 @@
         submissionQueue.submit();
 
         completionQueue.ioUringWaitCqe();
-        assertEquals(1, completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+        assertEquals(1, completionQueue.process(new IOUringCompletionQueueCallback() {
             @Override
             public void handle(int fd, int res, int flags, int op, int mask) {
                 assertEquals(inputString.length(), res);
@@ -73,7 +73,7 @@
         submissionQueue.submit();
 
         completionQueue.ioUringWaitCqe();
-        assertEquals(1, completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+        assertEquals(1, completionQueue.process(new IOUringCompletionQueueCallback() {
             @Override
             public void handle(int fd, int res, int flags, int op, int mask) {
                 assertEquals(inputString.length(), res);
@@ -105,7 +105,7 @@
             public void run() {
                 completionQueue.ioUringWaitCqe();
                 try {
-                    completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+                    completionQueue.process(new IOUringCompletionQueueCallback() {
                         @Override
                         public void handle(int fd, int res, int flags, int op, int mask) {
                             assertEquals(-62, res);
@@ -153,7 +153,7 @@
         }.start();
 
         completionQueue.ioUringWaitCqe();
-        assertEquals(1, completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+        assertEquals(1, completionQueue.process(new IOUringCompletionQueueCallback() {
             @Override
             public void handle(int fd, int res, int flags, int op, int mask) {
                 assertEquals(1, res);
@@ -184,7 +184,7 @@
             @Override
             public void run() {
                 completionQueue.ioUringWaitCqe();
-                assertEquals(1, completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+                assertEquals(1, completionQueue.process(new IOUringCompletionQueueCallback() {
                     @Override
                     public void handle(int fd, int res, int flags, int op, int mask) {
                         assertEquals(1, res);
@@ -234,8 +234,8 @@
 
         final AtomicReference<AssertionError> errorRef = new AtomicReference<AssertionError>();
         Thread waitingCqe = new Thread() {
-            private final IOUringCompletionQueue.IOUringCompletionQueueCallback verifyCallback =
-                    new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+            private final IOUringCompletionQueueCallback verifyCallback =
+                    new IOUringCompletionQueueCallback() {
                 @Override
                 public void handle(int fd, int res, int flags, int op, int mask) {
                     if (op == Native.IORING_OP_POLL_ADD) {
@@ -283,7 +283,7 @@
                 submissionQueue.addWrite(-1, -1, -1, -1, i);
                 assertEquals(1, submissionQueue.submitAndWait());
                 final int expectedData = i;
-                assertEquals(1, completionQueue.process(new IOUringCompletionQueue.IOUringCompletionQueueCallback() {
+                assertEquals(1, completionQueue.process(new IOUringCompletionQueueCallback() {
                     @Override
                     public void handle(int fd, int res, int flags, int op, int data) {
                         assertEquals(-1, fd);
Index: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java	(revision ea5a539bc60ad29d40895844051c61e482338cc0)
+++ transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java	(date 1601286542685)
@@ -20,7 +20,6 @@
 import io.netty.channel.unix.Errors;
 import io.netty.channel.unix.FileDescriptor;
 import io.netty.channel.unix.IovArray;
-import io.netty.channel.uring.IOUringCompletionQueue.IOUringCompletionQueueCallback;
 import io.netty.util.collection.IntObjectHashMap;
 import io.netty.util.collection.IntObjectMap;
 import io.netty.util.concurrent.RejectedExecutionHandler;
Index: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java	(revision ea5a539bc60ad29d40895844051c61e482338cc0)
+++ transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java	(date 1601286707628)
@@ -19,6 +19,7 @@
 import io.netty.util.internal.logging.InternalLogger;
 import io.netty.util.internal.logging.InternalLoggerFactory;
 
+import static io.netty.channel.uring.UserData.encode;
 import static java.lang.Math.max;
 import static java.lang.Math.min;
 
@@ -126,7 +127,7 @@
         PlatformDependent.putLong(sqe + SQE_ADDRESS_FIELD, bufferAddress);
         PlatformDependent.putInt(sqe + SQE_LEN_FIELD, length);
         PlatformDependent.putInt(sqe + SQE_RW_FLAGS_FIELD, rwFlags);
-        long userData = convertToUserData(fd, op, data);
+        long userData = encode(fd, op, data);
         PlatformDependent.putLong(sqe + SQE_USER_DATA_FIELD, userData);
 
         logger.trace("UserDataField: {}", userData);
@@ -180,7 +181,7 @@
     //fill the address which is associated with server poll link user_data
     boolean addPollRemove(int fd, int pollMask, int extraData) {
         return enqueueSqe(Native.IORING_OP_POLL_REMOVE, 0, fd,
-                convertToUserData(fd, Native.IORING_OP_POLL_ADD, pollMask), 0, 0, extraData);
+                          encode(fd, Native.IORING_OP_POLL_ADD, pollMask), 0, 0, extraData);
     }
 
     boolean addConnect(int fd, long socketAddress, long socketAddressLength, int extraData) {
@@ -242,13 +243,6 @@
         PlatformDependent.putLong(timeoutMemoryAddress + KERNEL_TIMESPEC_TV_NSEC_FIELD, nanoSeconds);
     }
 
-    private static long convertToUserData(int fd, int op, int data) {
-        assert op <= Short.MAX_VALUE;
-        assert data <= Short.MAX_VALUE;
-        int opMask = op << 16 | (((short) data) & 0xFFFF);
-        return ((long) fd) << 32 | opMask & 0xFFFFFFFFL;
-    }
-
     public long count() {
         return tail - head;
     }
Index: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueueCallback.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueueCallback.java	(date 1601286822330)
+++ transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueueCallback.java	(date 1601286822330)
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+interface IOUringCompletionQueueCallback {
+    /**
+     * Called for a completion event that was put into the {@link IOUringCompletionQueue}.
+     */
+    void handle(int fd, int res, int flags, int op, int data);
+}
Index: transport-native-io_uring/src/test/java/io/netty/channel/uring/UserDataTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/test/java/io/netty/channel/uring/UserDataTest.java	(date 1601286707626)
+++ transport-native-io_uring/src/test/java/io/netty/channel/uring/UserDataTest.java	(date 1601286707626)
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class UserDataTest {
+    @Test
+    public void testUserData() {
+        // Ensure userdata works with negative and positive values
+        for (int fd : new int[] { 0, 1, 10, Short.MAX_VALUE, Integer.MAX_VALUE }) {
+            for (int op = 0; op < 20; op++) {
+                for (int data = Short.MIN_VALUE; data <= Short.MAX_VALUE; data++) {
+                    final int expectedFd = fd;
+                    final int expectedOp = op;
+                    final int expectedData = data;
+                    long udata = UserData.encode(expectedFd, expectedOp, expectedData);
+                    UserData.decode(0, 0, udata, new IOUringCompletionQueueCallback() {
+                        @Override
+                        public void handle(int actualFd, int res, int flags, int actualOp, int actualData) {
+                            assertEquals(expectedFd, actualFd);
+                            assertEquals(expectedOp, actualOp);
+                            assertEquals(expectedData, actualData);
+                        }
+                    });
+                }
+            }
+        }
+    }
+}
\ No newline at end of file
Index: transport-native-io_uring/src/main/java/io/netty/channel/uring/UserData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/main/java/io/netty/channel/uring/UserData.java	(date 1601286707623)
+++ transport-native-io_uring/src/main/java/io/netty/channel/uring/UserData.java	(date 1601286707623)
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+final class UserData {
+    private UserData() {
+    }
+
+    static long encode(int fd, int op, int data) {
+        assert op <= Short.MAX_VALUE;
+        assert data <= Short.MAX_VALUE;
+        return (long) data << 48 | (long) op << 32 | fd;
+    }
+
+    static void decode(int res, int flags, long udata, IOUringCompletionQueueCallback callback) {
+        int fd = (int) (udata & 0xFFFFFFFFL);
+        int op = (short) ((udata >>>= 32) & 0xFFFFL);
+        int data = (short) (udata >>> 16);
+        callback.handle(fd, res, flags, op, data);
+    }
+}
Index: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java	(revision ea5a539bc60ad29d40895844051c61e482338cc0)
+++ transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java	(date 1601286707630)
@@ -17,6 +17,8 @@
 
 import io.netty.util.internal.PlatformDependent;
 
+import static io.netty.channel.uring.UserData.decode;
+
 /**
  * Completion queue implementation for io_uring.
  */
@@ -83,24 +85,13 @@
             ringHead++;
             PlatformDependent.putIntOrdered(kHeadAddress, ringHead);
 
-            int fd = (int) (udata >>> 32);
-            int opMask = (int) udata;
-            int op = opMask >>> 16;
-            int data = (short) opMask;
-
             i++;
-            callback.handle(fd, res, flags, op, data);
+
+            decode(res, flags, udata, callback);
         }
         return i;
     }
 
-    interface IOUringCompletionQueueCallback {
-        /**
-         * Called for a completion event that was put into the {@link IOUringCompletionQueue}.
-         */
-        void handle(int fd, int res, int flags, int op, int data);
-    }
-
     /**
      * Block until there is at least one completion ready to be processed.
      */
```

Reply:
@chrisvest makes sense... Let me apply this one .

Reply:
`return (long) data << 48 | (long) op << 32 | fd; `@chrisvest why is data a short? :)

Reply:
@1Jo1 because a short is all we can store in there... we only have 16 bytes.

Reply:
it's a bit confusing, because the `data` is actually a int

Reply:
its an int just to make it easier to pass around (but as you may have noticed we have an assert in there to ensure it is never bigger than a short). Otherwise you will need to always add `(short) ` when passing it around

Reply:
ok, I would suggest to change data type parameter to short, it would make more sense

Reply:
ah that makes sense 👍

Reply:
```java
  +        return (long) data << 48 | (long) op << 32 | fd;
```
May be best to mask `fd` here since we are otherwise implicitly relying on it being non-negative.

I tend to agree with @1Jo1 if we are going to interpret the 16bit vals as signed short then would be better for those args of the `handle` method to be shorts.